editor/src/components/canvas/canvas-wrapper-component.tsx

```diff
@@ -93,11 +119,296 @@ export const CanvasWrapperComponent = React.memo(() => {
+        x: selectionArea.x,
+        y: selectionArea.y,
+      }),
+    ).canvasPositionRounded
+    const bottomRight = windowToCanvasCoordinates(
+      canvasScale,
+      canvasOffset,
+      windowPoint({
+        x: selectionArea.x + selectionArea.width,
+        y: selectionArea.y + selectionArea.height,
+      }),
+    ).canvasPositionRounded
+
+    const rect = canvasRectangle({
+      x: topLeft.x,
+      y: topLeft.y,
+      width: bottomRight.x - topLeft.x,
+      height: bottomRight.y - topLeft.y,
+    })
+    for (const element of Object.values(metadata)) {
+      const isChildOfSceneRoot = MetadataUtils.isProbablyScene(
+        metadata,
+        EP.nthParentPath(element.elementPath, 3),
+      )
+      const isValidTarget =
+        isChildOfSceneRoot || EP.isStoryboardPath(EP.parentPath(element.elementPath))
+      const frame = element.globalFrame
+      if (
+        frame != null &&
+        isFiniteRectangle(frame) &&
+        rectanglesOverlap(frame, rect) &&
+        isValidTarget
+      ) {
+        res.push({
+          path: element.elementPath,
+          type: isChildOfSceneRoot
+            ? 'scene-child'
+            : MetadataUtils.isProbablyScene(metadata, element.elementPath)
+            ? 'scene'
+            : 'storyboard-child',
+          fullyCovered: rectangleContainsRectangle(rect, frame),
+        })
+      }
+    }
+
+    const thereAreStoryboardChildren = res.some((other) => other.type === 'storyboard-child')
+    res = res.filter((e) => {
+      // if the element is a schene child and there are storyboard children, skip it
+      if (e.type === 'scene-child' && thereAreStoryboardChildren) {
+        return false
+      }
+      // if the element is a scene, and the scene is not fully covered skip the scene
+      if (e.type === 'scene' && !e.fullyCovered) {
+        return false
+      }
+      // if a scene is fully covered, select just the scene and omit its children
+      if (e.type === 'scene-child') {
+        const parentScene = res.find(
+          (other) => other.type === 'scene' && EP.isDescendantOf(e.path, other.path),
+        )
+        if (parentScene != null && parentScene.fullyCovered) {
+          return false
+        }
+      }
+      return true
+    })
+
+    setElementsUnderSelectionArea((old) => {
+      const paths = res.map((e) => e.path)
+      if (old.length !== res.length) {
+        return paths
+      }
+      if (!old.every((a) => paths.some((b) => EP.pathsEqual(a, b)))) {
+        return paths
+      }
+      return old
+    })
+  }, [selectionArea, metadata, canvasScale, canvasOffset, mousePoint, selectionAreaStart])
+
+  React.useEffect(() => {
+    if (mode.type !== 'select') {
+      setSelectionAreaStart(null)
+    }
+  }, [mode.type])
+
+  React.useEffect(() => {
+    if (elementsUnderSelectionArea == null) {
+      return
+    }
+    dispatch([
+      setHighlightedViews(elementsUnderSelectionArea),
+      setHoveredViews(elementsUnderSelectionArea),
+    ])
```

Comment:
I'd be wary of dispatching actions in `useEffect` call, because I'm afraid this might be hard to debug later on. Can this be rephrased so that the action dispatch is tied to a more "discrete" action, such as a mouse move handler firing?